### PR TITLE
Add Features - Token Conversion, Popout Chat, Show Hidden Items

### PR DESCRIPTION
**Pull request by tootster**
Apologies for the disgusting commit history I couldn't work out how to exclude those files so have created a new pull request with a few other things, 

Ive fixed the issue with hidden items and it runs correctly on startup now.

Ive also added Token Conversion which converts tokens to a dollar value with the ability to also change the exchange rate so people can use their local currency. (The base conversion rate to USD is the average price per token in USD)

Also added a popout chat window, this doesn't let you actually chat (cant work out how to do that) But it looks identical to the chat window on the site and still shows you the messages received.

### CODE REVIEW

**Comment by f3rked** [comment]
Thanks for adding - I'll review tomorrow.  For futures, I'd prefer smaller PRs so it's easier to review - like keeping diff features in diff PRs if possible.  It's fine for this one though.

Also does the popout chat do anything different than just going to https://www.fishtank.live/chat

**Comment by tootster** [comment]
Didnt even know that chat link was a thing lol, and no it doesnt, does less actually because you cant use it to chat it just displays the chat messages.

**Comment by f3rked** [comment]
> Didnt even know that chat link was a thing lol, and no it doesnt, does less actually because you cant use it to chat it just displays the chat messages.

Yea a lot of people don't seem to know about it.  Might be a worthwhile addition to have a link that opens it in a new tab.

File: src/modules/config.js [hunk 1]
```diff
@@ -126,7 +131,7 @@ const Config = () => {
                 options: [
                   {
                     type: "number",
-                    valid: "number",
+                    valid: "integer",
```
Comment (f3rked) [inline src/modules/config.js:134]: why is this changed? is this broken?

File: src/main.js [hunk 2]
```diff
@@ -27,8 +29,10 @@ import "./styles/styles.scss";
 
   const enableDimMode =
     config.get("enableDimMode") && config.get("enablePlugin");
-  toggleDimMode(enableDimMode);
 
+  toggleDimMode(enableDimMode);
+  toggleHiddenItems(config.get("showHiddenItems") && config.get("enablePlugin"));
+  toggleTokenConversion(config.get("convertTokenValues") && config.get("enablePlugin"));
```
Comment (f3rked) [inline src/main.js:35]: does it not work on refresh if you put this in the startMaejokTools function i mentioned?

File: src/modules/functions.js [hunk 3]
```diff
@@ -323,6 +323,328 @@ export const displayUserNameOverlay = () => {
   playerHeaderTarget.insertAdjacentElement("beforebegin", userOverlayContainer);
 };
 
+export const toggleTokenConversion = (toggle) => {
+  const tokenToUsdRate = 0.0828; // Conversion rate from tokens to USD
```
Comment (f3rked) [inline src/modules/functions.js:327]: what is this based on? the usd -> token conversion is different based on how many tokens you buy at once.  i haven't tested this yet, but if i decide to include it, i'd like to include a disclaimer about what rate it's based on and that it won't necessarily reflect the amount paid for those tokens

File: src/modules/config.js [hunk 4]
```diff
@@ -9,6 +9,11 @@ const Config = () => {
     enablePlugin: true,
     enableDimMode: false,
 
+    convertTokenValues: false,
+    usdExchangeRate: 1.00,
```
Comment (f3rked) [inline src/modules/config.js:13]: please put this in `constants.js`

File: src/modules/functions.js [hunk 5]
```diff
@@ -323,6 +323,328 @@ export const displayUserNameOverlay = () => {
...
+  const cfg = config.get();
```
Comment (f3rked) [inline src/modules/functions.js:328]: import the value from constants after you move it there - no need to store all settings in memory again

File: src/modules/functions.js [hunk 6]
```diff
@@ -323,6 +323,328 @@ export const displayUserNameOverlay = () => {
...
+          convertTokensToLocalCurrency(element);
+        } else {
+          revertToOriginalTokens(element);
+        }
+      });
+    });
+  };
+
+  // MutationObserver to process newly added elements, including tokens in dynamically created modals
+  const observer = new MutationObserver((mutations) => {
```
Comment (f3rked) [inline src/modules/functions.js:433]: please use the existing modal observer in `observers.js`

File: src/modules/config.js [hunk 7]
```diff
@@ -364,6 +428,9 @@ const Config = () => {
         },
       },
 
+
```
Comment (f3rked) [inline src/modules/config.js:431]: unnecessary newlines.  if you're using vscode and install a code formatter extension like `esbenp.prettier-vscode` it will use the .prettierrc in the repo to automatically correct stuff like this and help keep things more consistent.

**Review by f3rked — CHANGES_REQUESTED** [review]
if you want to put the hidden item highlighting in a separate PR, i can go ahead and merge that in, but i have some comments on the new stuff here.

i'd like the popout chat removed given fishtank.live/chat exists.  

also please remove the changes to the package.json and lock file.

**Comment by tootster** [comment]
Added a new commit with the changes you wanted, let me know if thats all good :)

File: src/modules/functions.js (hunk 6 above)
Comment (tootster) [reply src/modules/functions.js]: Moved Observer to Observers.js (Cant use existing as it needs to access items in both modals and the main body)

File: src/modules/config.js (hunk 7 above)
Comment (tootster) [reply src/modules/config.js]: Cleaned up whitespace correctly hopefully (I installed the extension but I don't know if the config is right)

File: src/modules/config.js (hunk 4 above)
Comment (tootster) [reply src/modules/config.js:13]: Not sure about moving those values to constants.js they need to be dynamically changed based on what the users saved.

File: src/modules/functions.js (hunk 3 above)
Comment (tootster) [reply src/modules/functions.js]: Its an average of all the different rates the site offers, I added an explanation about it and an option to change it

File: src/modules/functions.js [hunk 8]
```diff
@@ -323,6 +323,192 @@ export const displayUserNameOverlay = () => {
...
+        }
+      });
+    });
+  };
+
+  // Initial processing
+  processElements();
+};
+
+export const togglePopoutChatButton = (toggle) => {
```
Comment (f3rked) [inline src/modules/functions.js:433]: left a [comment](https://github.com/f3rked/maej3rked-tools/pull/29#pullrequestreview-2435151049) about removing this

File: src/modules/config.js (hunk 4 above)
Comment (f3rked) [reply src/modules/config.js:13]: ah i see

File: package.json [hunk 9]
```diff
@@ -61,4 +61,4 @@
     "webpack": "^5.89.0",
     "webpack-cli": "^5.1.4"
   }
-}
+}
```
Comment (f3rked) [inline package.json:64]: remove package.json and lockfile changes

File: src/modules/observers.js [hunk 10]
```diff
@@ -300,6 +301,49 @@ const observers = {
       observers.body?.disconnect();
     },
   },
+
+  tokens: {
```
Comment (f3rked) [inline src/modules/observers.js:305]: are you trying to target when modals show up? if so that already exists [here](https://github.com/f3rked/maej3rked-tools/blob/6bda06936c5f7b7c21802578dca0ef515bdac096/src/modules/observers.js#L92).  if any of these appear outside of a modal, there is already a body observer [here](https://github.com/f3rked/maej3rked-tools/blob/6bda06936c5f7b7c21802578dca0ef515bdac096/src/modules/observers.js#L262).  

observing dom changes recursively isn't necessarily cheap. to explain my concern:
a.) multiple dom observers for the entire body is unnecessary
b.) callback code should run in as targeted a way as possible rather than wasting cpu cycles iterating through every dom change that happens

i realize the observers code here isn't the most efficient and needs to be refactored, but i don't want to make it worse.

**Review by f3rked — CHANGES_REQUESTED** [review]

File: src/modules/functions.js (hunk 8 above)
Comment (tootster) [reply src/modules/functions.js:433]: Ive changed this to just add a button at the top right of the chat element that opens a link to fishtank.live/chat is that okay or you want that removed too?

**Comment by tootster** [comment]
The current observer setup couldn’t detect all the elements needed for token conversion because it required "subtree: true". To solve this, I added a second observer within the main modal observer. This second observer only activates when a modal is open.

The second observer uses "subtree: true" so it can detect dynamic updates inside the modal, like in the fish toys modal, where content changes when switching tabs (e.g., bigtoys, wartoys). This observer handles token conversions as new elements load or change within the modal. When the modal closes, this second observer stops automatically to save resources.

Are you okay with this approach?

File: package.json (hunk 9 above)
Comment (tootster) [reply package.json]: Fixed this, not sure why it was showing this as a diff it was exactly the same...

File: src/modules/functions.js (hunk 8 above)
Comment (f3rked) [reply src/modules/functions.js:432]: my bad - probably should have waited longer after i woke up to review.

that sounds good.  i'll test later today to see what it looks like.

File: src/modules/observers.js (hunk 10 above)
Comment (tootster) [reply src/modules/observers.js]: The current observer setup couldn’t detect all the elements needed for token conversion because it required "subtree: true". To solve this, I added a second observer within the main modal observer. This second observer only activates when a modal is open.

The second observer uses "subtree: true" so it can detect dynamic updates inside the modal, like in the fish toys modal, where content changes when switching tabs (e.g., bigtoys, wartoys). This observer handles token conversions as new elements load or change within the modal. When the modal closes, this second observer stops automatically to save resources.

Are you okay with this approach?

**Comment by f3rked** [comment]
my bad started testing but couldn't focus with all the shit going on today.  i'll finish up tomorrow and merge hopefully.
